flink-python/src/main/java/org/apache/flink/python/util/TarUtils.java

```diff
@@ -0,0 +1,98 @@
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+/** Utils used to extract tar files. */
+@Internal
+public class TarUtils {
+    public static void unTar(String inFilePath, String targetDirPath)
+            throws IOException, InterruptedException {
+        File targetDir = new File(targetDirPath);
+        if (!targetDir.mkdirs()) {
+            if (!targetDir.isDirectory()) {
+                throw new IOException("Mkdirs failed to create " + targetDir);
+            }
+        }
+        boolean gzipped = inFilePath.endsWith("gz");
+        if (DecompressUtils.isUnix()) {
+            unTarUsingTar(inFilePath, targetDirPath, gzipped);
+        } else {
+            unTarUsingJava(inFilePath, targetDirPath, gzipped);
+        }
+    }
+
+    // Copy and simplify from hadoop-common package that is used in YARN
+    // See https://github.com/apache/hadoop/blob/7f93349ee74da5f35276b7535781714501ab2457/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileUtil.java
+    private static void unTarUsingTar(String inFilePath, String targetDirPath, boolean gzipped)
+            throws IOException, InterruptedException {
+        inFilePath = makeSecureShellPath(inFilePath);
+        targetDirPath = makeSecureShellPath(targetDirPath);
+        String untarCommand =
+                gzipped
+                        ? String.format(
+                                "gzip -dc '%s' | (cd '%s' && tar -xf -)", inFilePath, targetDirPath)
+                        : String.format("cd '%s' && tar -xf '%s'", targetDirPath, inFilePath);
+        Process process = new ProcessBuilder("bash", "-c", untarCommand).start();
+        int exitCode = process.waitFor();
+        if (exitCode != 0) {
+            throw new IOException(
+                    "Error untarring file "
+                            + inFilePath
+                            + ". Tar process exited with exit code "
+                            + exitCode);
+        }
+    }
+
+    // Follow the pattern suggested in
+    // https://commons.apache.org/proper/commons-compress/examples.html
+    private static void unTarUsingJava(String inFilePath, String targetDirPath, boolean gzipped)
+            throws IOException {
+        try (InputStream fi = Files.newInputStream(Paths.get(inFilePath));
+                InputStream bi = new BufferedInputStream(fi);
+                ArchiveInputStream ai =
+                        new TarArchiveInputStream(
+                                gzipped ? new GzipCompressorInputStream(bi) : bi)) {
+            ArchiveEntry entry;
+            while ((entry = ai.getNextEntry()) != null) {
+                File f = new File(targetDirPath, entry.getName());
+                if (entry.isDirectory()) {
+                    if (!f.isDirectory() && !f.mkdirs()) {
+                        throw new IOException("failed to create directory " + f);
+                    }
+                } else {
```

Comment:
It also considered link and hard link in https://github.com/apache/hadoop/blob/7f93349ee74da5f35276b7535781714501ab2457/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileUtil.java#L991. Should we also consider that?